### DISCUSSION TITLE
Multi-node MQTT server

### DISCUSSION THREAD

**Discussion by raj-cg** [mochi-mqtt/server#174] [Q&A]
We are trying to implement a high-availability MQTT service using mochi, If a persistence layer of Redis is added then multiple nodes of mochi should be able to connect to the same Redis backend and retain session data in Redis, 

Also is there a need for sticky sessions when working with load balancers? 

Is there any additional work required to create a HA setup?
Please let me know if I missed anything here.

**Comment by dhruvjain99** [comment]
How do you plan to pass messages around for scenarios in which clients are on different nodes for which the message needs to be routed?

**Reply by RohanDoshi21** [reply]
What could the different approaches be?

I was thinking of
1. Broadcasting all publish messages to each node. ( filter out based upon if client is connected to me, avoid loops but not further broadcasting if processing a broadcasted message )

2. Using Redis Pub-Sub. ( Each node will be a redis client subscriber ), but this doesn't guarantee message delivery.

3. Using Redis streams, and clearing out old messages periodically / clearing out messages after grp ack is received. 

Stuck at making a design decision, trying to create a clustered setup, for a college project! 
Any thoughts upon this @dhruvjain99 @raj-cg @wind-c

**Comment by wind-c** [comment]
Comqtt's cluster capability was developed based on mochi-co/mqtt v2.0.
It uses gossip for node discovery and raft for data(subscribe/unsubscribe) consistency synchronization. It uses redis to store inflight, retain messages and subscriptions across nodes.Publish messages are routed to the corresponding node according to the subscription table. [For reference](https://github.com/wind-c/comqtt/blob/main/cmd/cluster/README.md).

**Comment by torntrousers** [comment]
> Comqtt's cluster capability was developed based on mochi-co/mqtt v2.0. It uses gossip for node discovery and raft for data(subscribe/unsubscribe) consistency synchronization. It uses redis to store inflight, retain messages and subscriptions across nodes.Publish messages are routed to the corresponding node according to the subscription table. [For reference](https://github.com/wind-c/comqtt/blob/main/cmd/cluster/README.md).

@wind-c - sounds really interesting - why not come up with some PR proposals to get that merged in here?
